[IMP] rest_log: handle odoo.http.Response objects in logs

SilvioC2C opened the pull request:
Handle `odoo.http.Response` objects when creating logs.

simahawk committed:
rest_log: inject log entry URL only for dict response

simahawk commented:
> Handle `odoo.http.Response` objects when creating logs.

Can you please explain what you are trying to achieve?
Also, rebase and drop the change on README.md.

SilvioC2C commented:
@simahawk 

Some endpoints will not return a dict as expected by the ``_dispatch_with_db_logging`` method (and the other methods called within), but an object of type ``odoo.http.Response``.
When that happens, the instance fails to handle it while creating logs (ie: tries to call ``json.dumps()`` on such object, which is not allowed).
This patch should fix the issue.

jcoux reviewed (approved):
Maybe you can just add a comment to explain in case of returned an attachment (PDF by example), the service will return directly a response and not a dict.

simahawk commented on rest_log/components/service.py:
```diff
@@ -68,7 +68,7 @@ def _dispatch_with_db_logging(self, method_name, *args, params=None):
             self.env, request, method_name, *args, params=params, result=result
         )
-        if log_entry:
+        if log_entry and isinstance(result, dict):
```
this has been fixed in v14 already (I think we lost it in v13) https://github.com/OCA/rest-framework/commit/4e78558a1ec4fdd77df35017d2f09f2b68d337d9
Please cherry-pick this commit instead.

simahawk commented on rest_log/components/service.py:
```diff
@@ -123,6 +123,12 @@ def _log_call_in_db_values(self, _request, *args, params=None, **kw):
         result = kw.get("result")
+        if isinstance(result, Response):
+            status_code = result.status_code
+            result = {"status_code": status_code}
```
Maybe content length and filename from the header if they are available?
Also, please, add a test for this :pray:

simahawk commented:
> @simahawk
> 
> Some endpoints will not return a dict as expected by the `_dispatch_with_db_logging` method (and the other methods called within), but an object of type `odoo.http.Response`. When that happens, the instance fails to handle it while creating logs (ie: tries to call `json.dumps()` on such object, which is not allowed). This patch should fix the issue.

This should be the description of the commit :wink: 
You can also shorten the message by removing "in logs" which is redundant for "rest_log".

simahawk committed:
rest_log: add sanitize params hook

Provide a hook to customize stored params.
At the same time, assume passwords must be always cleaned up.

SilvioC2C commented:
@simahawk 
Added headers to `result`, also added unit tests.
Mind taking a look again? :sweat_smile:

simahawk commented on rest_log/components/service.py:
```diff
@@ -123,6 +123,15 @@ def _log_call_in_db_values(self, _request, *args, params=None, **kw):
+        # handle it properly, without the assumption that ``result`` is a dict.
+        if isinstance(result, Response):
+            status_code = result.status_code
+            result = {"status": status_code, "headers": dict(result.headers)}
```
I think you must sanitize headers here like is done above. Maybe move it to a new method.

Can you also cherry-pick this fix? https://github.com/OCA/rest-framework/pull/301

SilvioC2C replied:
@simahawk done

SilvioC2C committed:
[FIX] rest_log: fix params management

SilvioC2C committed:
[IMP] rest_log: refactor headers management

SilvioC2C committed:
[IMP] rest_log: handle odoo.http.Response objects in logs

Some endpoints will not return a dict as expected by the ``_dispatch_with_db_logging`` method (and the other methods called within), but an object of type ``odoo.http.Response``.
When that happens, the instance fails to handle it while creating logs (ie: tries to call ``json.dumps()`` on such object, which is not allowed).
This patch should fix the issue.

simahawk commented:
/ocabot merge minor
